File: week-03_05/tp/tp_C/tp2/ping_pong_2.c
#include "mpi.h"
#include <stdlib.h>
#include <stdio.h>

int main(int argc, char *argv[]) {
  int rank,iter;
  int nb_values=1000;
  int tag=99;
  double values[nb_values];
  MPI_Status status;
  double time_begin,time_end;

  /* TODO : Get rank */

  if (rank == 0) {
    for (iter = 0; iter<nb_values; iter++)
      values[iter] = rand() / (RAND_MAX + 1.);
    /* TODO : Get starting time -> time_begin */
    
    /* TODO : Process 0 send and receive values with process 1 */

    /* TODO : Get ending time -> time_end */

    printf("Me, process 0, sent and received %d values"
           "(last = %g) from process 1 in %f seconds.\n",
           nb_values, values[nb_values-1], time_end-time_begin);
  else {
    /* TODO : Process 1 receive values and send them with process 0 */

  }

  /* TODO : Finalize MPI */
  
  return 0;
}

File: week-03_05/tp/tp/tp2/ping_pong_2.f90
program ping_pong_2
  use, intrinsic :: iso_fortran_env, only : dp => real64
  USE MPI_F08
  implicit none

  TYPE(MPI_Status)                     :: status
  integer, parameter                   :: nb_values=1000,tag=99
  integer                              :: rank
  real(kind=dp)                        :: time_begin,time_end
  real(kind=dp), dimension(nb_values)  :: values
  TYPE(MPI_Datatype)                   :: typedp

  ! TODO Get rank -> rank

  if (rank == 0) then
    call random_number(values)
    ! TODO: Start timing -> time_begin

    ! TODO: Process 0 send values to and receive from process 1

    ! TODO: End timing -> time_end

    print ('("Me, process 0, sent and received ",i5, &
        & " values (last = ",f4.2,") from process 1", &
        & " in ",f8.6," seconds.")'), &
          nb_values,values(nb_values),time_end-time_begin

    ! TODO: Process 1 receive values from and send to process 0

  end if

! TODO: Finalize MPI

end program ping_pong_2
